---
title: "Basic Data Visualization"
subtitle: "Amplicon data visualization and linear models"
author: "Judy-Malas"
date: "12/15/2021"
output: 
  html_document: 
    theme: readable
---

This workflow will use tidyverse functions and ggplot to make visualizations such as boxplots, heatmaps, and histograms. This is meant as a first pass at seeing what the data look like. 

We also try our hand at linear models, however, these are likely not the best statistical tools for 16s data. 
Step 4; ordination, and Step 5; differential expression, are probably better tools to get some statistical power behind our results. 

Before Starting: 

* clear the global environment by clicking the broom AFTER making sure you have saved all the updated phyloseq objects from step 2


## Loading packages & data 
If phyloseq is not installed, un-comment the commented section. 
```{r load-packages, message = FALSE}
library(tidyverse)
library(broom)
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("phyloseq")
library(phyloseq)
library(forcats)
library(tidymodels)
library(RColorBrewer)
```

This is the main dataset used -- but I will also run some chunks with the other phyloseq objects created in step 2, depending on what I would like to do
```{r read_phyloseq_object}
full_16s_dataset <- readRDS("../data/live_samples_objects/ps_all.step2")
```

I don't need these in this step, so I'll comment it out for now and uncomment one I need one of these transformed phyloseq objects. 
```{r import-all-objects}

# live samples 

# ps_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.step2") 
# 
# ps_all.rmout <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rmout") #outlier removed
# 
# vst_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/vst_all.step2")
# 
# ps_all.rl <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rl") 
# 
# ps_all.prop <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.prop") 
# 
# 
# 
# 
#  #killed samples 
# 
# ps_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.step2")
# 
# ps_killed.rmout <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rmout")
# 
# vst_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.step2")
# 
# ps_killed.rl <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rl")
# 
# ps_killed.prop <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.prop")


```

In order to analyze this data with the tidyverse functions, we need to convert the phylsoeq object to a regular data frame. 
First we use the `transform_sample_counts` function to change our absolute counts into relative count, which are more useful for visualization than absolute counts. 
Alternatively, we can skip this step and use use any of the already transformed phyloseq objects the following step with the function "ps_melt"
```{r define-calc-prop}

calc_prop <- function(x){x / sum(x)}
# 
full_16s_dataset_relative <- transform_sample_counts(full_16s_dataset, calc_prop)
```

We can then use the function `psmelt` which will combine all the elements of the object, except the phylogentic tree, into a single data frame.
Here I made two dataframes: one for the full dataset with absolute counts, and one for the full dataset with relative counts. 
```{r create-dataframe-with-psmelt}

full_16s_df_rel <- psmelt(full_16s_dataset_relative)

full_16s_df_abs <- psmelt(full_16s_dataset)

```


##Tidying data 
I don't think I need this anymore because I already re-leveled the spikes in step-2
```{r relevel-spikes, results = 'hide'}
# as.factor(full_16s_df_rel$spike)
# 
# full_16s_df_rel <- full_16s_df_rel %>% 
#   mutate(spike = fct_relevel(spike, "Control ", "Antibiotics", "Fe(OH)3", "Na2SO4"))

```


## Exploratory Visualization

This chunk will apply a theme to all of our plots to make them look pretty. 
```{r set-ggplot2-theme}
theme_set(theme_bw())
```


Let's check out the phyla present in the data. (Absolute counts, but it really doesn't matter if we use relative or absolute here)
```{r group-by-phyla-desc}

 full_16s_df_abs %>%
  group_by(Phylum) %>% 
   count() %>% 
   arrange(desc(n)) 

```



*Firmicutes*, *Proteobacteria*, and *Actinobacteriota* are the most abudant Phyla. 


```{r histograms-of-abundance}
ggplot(full_16s_df_rel, aes(Abundance)) +
  geom_histogram() +
  labs(x = "Relative Abundance")

```

From the plot above we can see that most of the samples have OTUs with very low relative abundances <002. Perhaps from this plot we can get an understanding of how much of the data gets cut off when we filter for different abundances.


The following chunk arranges the full dataset in descending order by Genus counts. 
```{r group-by-Genus-desc, echo = FALSE}
genus_desending <- full_16s_df_abs %>% 
   group_by(Genus) %>% 
   count() %>%  #count the number of observations in each Genus group 
   arrange(desc(n))  #arrange the counts in descending order

genus_desending

```
*Caproiciproducens, Incertae_Sedis*, and *Oxobacter* are the 3 most abundant genera. 



```{r plot-genus-counts}
  ggplot(genus_desending, aes(n)) +
  geom_histogram() + 
  theme_bw()
```
There are 127 identified Genuses, are most of them are observed <10000 times, with one outlier (*Caproiciproducens*). That being said, the vast majority of genera are NA. 

### Visualization plots of relative abudances throughout the dataset

#### Heatmaps
```{r heatmap-by-spike-phylum-level-T6}

full_16s_df_rel %>% 
  #filter(sample_time == c("T6", "T1")) %>% #filter for only T1 & T6 
  group_by(spike, Phylum, sample_time, Abundance, Genus) %>%  #select only these columns 
  summarise(mean_abundance = mean(Abundance)) %>%  #this line doesn't seem to be doing anything but the plot won't work without it.. 
  arrange(sample_time) %>% 
  ggplot(aes(x = spike, y = Phylum, fill = Abundance))+
    geom_tile() +
    facet_wrap(~sample_time) +
  theme(axis.text.x=element_text(angle = -90, hjust = 0))

```
This is a nice plot, I think I like the heat maps better than the stacked barcharts... We'll see how it looks at other taxanomy levels


```{r heatmaps-top-10-genera}

genera_top_10 = c("Anoxybacillus", "Caproiciproducens", "Cupriavidus", "Oxobacter", "Sporolactobacillus", "Turicella", "Vulcaniibacterium", "Unclassified", "Other < 2%")

full_16s_df_rel %>% 
  mutate(Genus = case_when(Abundance < 0.02 ~ "Other < 2%",
                           is.na(Genus) ~ "Unclassified",
                               TRUE ~ Genus)) %>%
  #filter(Genus == c("Anoxybacillus", "Caproiciproducens", "Cupriavidus", "Oxobacter", "Sporolactobacillus", "Turicella", "Vulcaniibacterium", #"Unclassified", "Other < 2%")) %>% 
  select(spike, Abundance, Phylum, Family, Genus, sample_time, bottle) %>%
  #summarise(mean_abundance = mean(Abundance)) %>% 
  group_by(spike,  Abundance, Phylum, Family, Genus, sample_time) %>% 
  summarise(mean_abundance = mean(Abundance)) %>% 
  ggplot(aes(x = spike, y = Genus, fill = mean_abundance))+
    geom_tile() +
    #facet_wrap(~sample_time) +
  theme(axis.text.x=element_text(angle = -90, hjust = 0))

```
This looks weird because different spikes have different concentrations of the top 10 genera in the entire dataset. Some may not have >2% of each of the most abundant taxa. Frustrating..


#### Boxplots
```{r boxplots-phyla}
full_16s_df_rel %>% filter(sample_time == c("T1","T6"), Abundance > 0) %>% 
  ggplot(aes(x= Phylum, y = Abundance, fill = Phylum)) +
  geom_boxplot() +
  scale_y_log10()+
  facet_grid(spike~ sample_time) +
  theme_bw() +
  labs(y = "Relative Abundance", x = " ")+
  theme(axis.text.x=element_text(angle = -90, hjust = 0)) 
  

```

```{r top-10-genera-boxplot}

full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) %>%
  ggplot(aes(x= Genus, y = Abundance, fill = Genus)) +
  geom_boxplot() +
  scale_y_log10()+
  facet_grid(spike~ sample_time) +
  theme_bw() +
  labs(y = "Relative Abundance", x = " ")+
  theme(axis.text.x=element_text(angle = -90, hjust = 0))+
  scale_color_viridis_d()


```

#### Barcharts 
```{r top-10-genera-barchart }
full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) %>%
  ggplot(aes(x = spike, y = Abundance, fill = Genus, color = Genus)) + 
  geom_bar(stat="identity")

full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) %>%
  ggplot(aes(x = sample_time, y = Abundance, fill = Genus, color = Genus)) + 
  geom_bar(stat = "Identity") +
  facet_wrap(~spike)

```

This is not working, dunno why but I think it was a bad plot either way
```{r look-at-triplicate-controls-over-time}

 cols = brewer.pal(9, "Spectral")

full_16s_df_rel %>% 
  select(spike, Abundance, Phylum, Family, Genus, sample_time, bottle) %>%
  filter(spike == "Control ") %>% 
  mutate(Genus = case_when(Abundance < 0.02 ~ "Other < 2%",
                           is.na(Genus) ~ "Unclassified",
                               TRUE ~ Genus)) %>% 
  #mutate(Genus = fct_relevel(Genus, "Anoxybacillus", "Caproiciproducens", "Cupriavidus", "Oxobacter", "Sporolactobacillus", "Turicella", "Vulcaniibacterium", #"Unclassified", "Other < 2%")) %>% 
  ggplot(aes(x = bottle, y = Abundance, fill= Genus, color = Genus)) +
  geom_bar(stat = "identity") +
  #facet_wrap(~sample_time)+
  scale_color_manual(values= cols) +
  scale_fill_manual(values= cols) +
  theme_bw()+
  labs(y = "Relative Abundance", x = " ")



```



#### Scatter Plots
```{r top-10-genera-points}

# x = spike
full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) %>%
  ggplot(aes(x = spike, y = Abundance, fill = Genus, color = Genus)) + 
  geom_point()

#x = sample time 
full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) %>%
  ggplot(aes(x = sample_time, y = Abundance, fill = Genus, color = Genus)) + 
  geom_point()

```
Not a very useful sccatter plot 


```{r}

ggplot(full_16s_df_rel, aes(spike, Abundance)) +
  geom_point()

```

```{r}
full_16s_df_rel %>% 
  filter(Genus == "Thermoanaerobacterium") %>% 
  ggplot(aes(spike, Abundance)) +
  geom_point()
```


```{r tumebacillus-scatterplot-by-spike}

full_16s_df_rel %>% 
  filter(Genus == "Tumebacillus") %>%
  ggplot(aes(spike, Abundance)) + 
  geom_point()

```
Tumebacillus exists in very low abundance in the control, <0.005, but is more abundant in the other spikes



### Statistical Analyses


#### Linear regression
```{r linear-reg-top-10-genera}
top_10_genera_rel = full_16s_df_rel %>% 
  filter(Genus == c("Caproiciproducens", "Incertae_Sedis", "Oxobacter", "Clostridium_sensu_stricto_12", "Halomonas", "Tumebacillus", "Aeribacillus", "Clostridium_sensu_stricto_7", "Brevibacillus", "Thermoanaerobacterium")) 

linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = top_10_genera_rel) %>% 
  tidy()
```
The linear regression is not significant for all fo the top 10 genera 

```{r linear-reg-Tume}
Tumebacillus = full_16s_df_rel %>% 
  filter(Genus == "Tumebacillus")

fit_tume =  linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Tumebacillus) 

linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Tumebacillus) %>% 
  tidy()

glance(fit_tume)$r.squared

```
There seems to be some linear correlation between spike and relative abundance of Tumebacillus

Let's visualize the differences 
```{r boxplots-for-tume}

full_16s_df_rel %>% 
  filter(Genus == "Tumebacillus") %>% 
  ggplot(aes(x = spike, y = Abundance)) +
  geom_boxplot() +
  facet_wrap(~sample_time)


```
It looks like the differences begin to appear before the spikes are added at T2 or T3.. 

Let's see if the linear correlation still holds if we only look at T5 and T6 
```{r just-after_spike}

Tumebacillus_after_spike = full_16s_df_rel %>% 
  filter(Genus == "Tumebacillus", sample_time == c("T5", "T6"))

fit_tume_3 =  linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Tumebacillus_after_spike) 


linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Tumebacillus_after_spike)  %>% 
  tidy()

glance(fit_tume_3)$r.squared

```


Let's make a linear model for Caproiciproducens, the most abudant genus
```{r linear-reg-Cap}

Caproiciproducens = full_16s_df_rel %>% 
  filter(Genus == "Caproiciproducens")

fit_caproic =  linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Caproiciproducens) 

linear_reg() %>% 
  set_engine("lm") %>% 
  fit(Abundance~ spike, data = Caproiciproducens) %>% 
  tidy()


glance(fit_caproic)$r.squared

```
The linear relationship is sigifianct but the R squared is really low.. Minor impact of spikes

