---
title: "ordination & diversity metrics"
author: "Judy-Malas"
date: "12/22/2021"
output: html_document
---

# Before Starting 

* Make sure your clear the global environment by clicking on the broom in R studio 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
### Load packages
```{r load-packages, message=FALSE, warning=FALSE}

library(ggplot2)
library(gridExtra)
library(phyloseq)
require("tidyverse")
library("dplyr")
library("rmarkdown")
library("vegan")
library("PoiClaClu")
library("doParallel")
library("plotly")
library("microbiome")
library("ggpubr")
library("viridis")
library("plyr")
library("magrittr")
library("scales")
library(microbiome) # data analysis and visualisation
library(phyloseq) # also the basis of data object. Data analysis and visualisation
library(RColorBrewer) # nice color options
library(ggpubr) # publication quality figures, based on ggplot2
library(DT) # interactive tables in html and markdown
library(dplyr) # data handling  
library(data.table)


```


```{r import-all-objects}

# live samples 

ps_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.step2") 

ps_all.rmout <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rmout") #outlier removed

vst_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/vst_all.step2")

ps_all.rl <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rl") 

ps_all.prop <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.prop") 

ps_all.5 <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.5") #low prevelance taxa removed, untransformed



 #killed samples 

ps_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.step2")

ps_killed.rmout <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rmout")

vst_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.step2")

ps_killed.rl <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rl")

ps_killed.prop <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.prop")


```



# Update Phyloseq Objects
#### update the metadata tables

Here we are adding a new "treatment" column to each of the phyloseq objects to reflect whether the microcosms have been spiked
```{r add_new_column, results='hide'}

ps_all.step2_samdata <- ps_all.step2@sam_data #extract the sample data 

new_sam_data <- ps_all.step2_samdata %>% 
  mutate(treatment =  #add a new column of data which specifies whether the samples have ACTUALLY been treated with a spike
           case_when(sample_time == "T1" ~ "None", 
                     sample_time == "T2" ~ "None", 
                    sample_time == "T3" ~ "None", 
                        sample_time == "T4" ~ "None", 
                        sample_time == "T5" ~ as.character(spike), 
                    sample_time == "T6" ~ as.character(spike))) 

new_sam_data <- new_sam_data %>% 
  mutate(treatment = ifelse(treatment == "Control", "None", treatment)) #change "Control" to "none" in the new treatment variable


as.factor(new_sam_data$treatment)

new_sam_data <- new_sam_data %>% 
  mutate(treatment = fct_relevel(treatment, "None", "Antibiotics", "Fe(OH)3", "Na2SO4"))
  


```

```{r add_new_column_killed_samdata, results='hide'}

ps_killed.step2_samdata <- ps_killed.step2@sam_data #extract the sample data 

new_sam_data_killed <- ps_killed.step2_samdata %>% 
  mutate(treatment =  #add a new column of data which specifies whether the samples have ACTUALLY been treated with a spike
           case_when(sample_time == "T1" ~ "None", 
                     sample_time == "T2" ~ "None", 
                    sample_time == "T3" ~ "None", 
                        sample_time == "T4" ~ "None", 
                        sample_time == "T5" ~ as.character(spike), 
                    sample_time == "T6" ~ as.character(spike), 
                    sample_time == "T7"~as.character(spike))) 

new_sam_data_killed <- new_sam_data_killed %>% 
  mutate(treatment = ifelse(treatment == "killed_control", "None", treatment)) #change "Control" to "none" in the new treatment variable


as.factor(new_sam_data_killed$treatment)

new_sam_data_killed <- new_sam_data_killed %>% 
  mutate(treatment = fct_relevel(treatment, "None", "Antibiotics", "Fe(OH)3", "Na2SO4"))

levels(new_sam_data_killed$treatment)

```

```{r update-physeq-objects}

ps_all.step2@sam_data = new_sam_data # assign the new dataframe to the phyloseq object
vst_all.step2@sam_data = new_sam_data
ps_killed.step2@sam_data = new_sam_data_killed
vst_killed.step2@sam_data = new_sam_data_killed


```


#### re-simplifying names
```{r rename-physeq-objects}

ps_all <- ps_all.step2
vst_all <- vst_all.step2
ps_killed <- ps_killed.step2
vst_killed <- vst_killed.step2

```



## Assigning plot colors
Here are some pallette optoins
```{r list-all-cb-friendly-palettes}

brewer.pal.info %>% 
  filter(colorblind == T)
  
```

```{r visualize-all-cb-friendly-palettes}
display.brewer.all(colorblindFriendly = T)
```


Ended up using colors in the 8-color plalette for color blindness: 

http://mkweb.bcgsc.ca/colorblind/img/colorblindness.palettes.v11.pdf

```{r updated-assign-colors}
exper_colors <- c("black", "#3DB7E9",  "#D55E00", "#F0E442")
```



# Distance Measures; Ordination plots

## Live Samples 

### Bray-Curtis
```{r nmds-bray-vst-all}

nmds_bray_vst_all <- ordinate(vst_all, method="NMDS", distance = "bray")

nmds_bray_vst_all_plot  <-
  plot_ordination(vst_all, nmds_bray_vst_all, color = "spike", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)

nmds_bray_vst_all_plot #print the plot


```

Here we're using the new treatment variable we made to make the plot
```{r nmds-bray-vst-all-treatment}

nmds_bray_vst_all_plot  <-
  plot_ordination(vst_all, nmds_bray_vst_all, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray_vst_all_plot +
  theme(legend.title = element_blank())+
  scale_shape_discrete(na.translate = F) +
  scale_color_manual(values = c("black", "#3DB7E9",  "#D55E00", "#F0E442"), na.value = NA, labels = c("None", "Antibiotics", "Fe(OH)3", "Na2SO4", " "))
#print the plot


ggsave("figures/nmds-bray-vst-live.jpeg", height = 6)

```




```{r bray-just-t6}

vst_all.t6 <- vst_all %>% 
  subset_samples(sample_time == "T6")

nmds_bray_vst_all.t6 <- ordinate(vst_all.t6, method="NMDS", distance = "bray")

nmds_bray_vst_all_plot.t6  <-
  plot_ordination(vst_all.t6, nmds_bray_vst_all.t6, color = "treatment", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw() +
    theme(legend.title = element_blank())+
    scale_shape_discrete(na.translate = F) +
    scale_color_manual(values = c("black", "#3DB7E9",  "#D55E00", "#F0E442"), na.value = NA, labels = c("None", "Antibiotics", "Fe(OH)3", "Na2SO4", " "))

nmds_bray_vst_all_plot.t6#print the plot

ggsave("figures/nmds_bray_live_t6_vst.jpeg", height = 6)



```


```{r bray-just-t6-t5}

vst_all.t5.6 <- vst_all %>% 
  subset_samples(sample_time %in% c("T6", "T5"))

nmds_bray_vst_all.t6.5 <- ordinate(vst_all.t5.6, method="NMDS", distance = "bray")

nmds_bray_vst_all_plot.t6.5  <-
  plot_ordination(vst_all.t5.6, nmds_bray_vst_all.t6.5, color = "treatment", title = "Bray-Curtis NMDS", shape = "sample_time")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5, )+ 
    theme_bw() +
    theme(legend.title = element_blank())+
    scale_shape_discrete(na.translate = F) +
  scale_shape_manual(values= c(7, 8))+
    scale_color_manual(values = c("black", "#3DB7E9",  "#D55E00", "#F0E442"), na.value = NA, labels = c("None", "Antibiotics", "Fe(OH)3", "Na2SO4", " "))

nmds_bray_vst_all_plot.t6.5 #print the plot

ggsave("figures/nmds_bray_live_t6and5_vst.jpeg", height = 6)



```

```{r just-t1-t4}

vst_all.t4 <- vst_all %>% 
  subset_samples(sample_time %in% c("T1", "T2", "T3", "T4"))

nmds_bray_vst_all.t4 <- ordinate(vst_all.t4, method="NMDS", distance = "bray")

cols_blue_4<- c('#FEDC56','#c7e9b4','#41b6c4','#253494')


nmds_bray_vst_all_plot.t4  <-
  plot_ordination(vst_all.t4, nmds_bray_vst_all.t4, color = "sample_time", title = "Bray-Curtis NMDS", shape = "sample_time")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5, )+ 
    theme_bw() +
    theme(legend.title = element_blank())+
    scale_shape_discrete(na.translate = F)+
    scale_color_manual(values = cols_blue_4)

nmds_bray_vst_all_plot.t4 #print the plot

ggsave("figures/nmds_bray_live_4_vst.jpeg", height = 6)

```



### Weighted Unifrac 
```{r pcoa-wunifrac-vst-all}

PCoA_wunifrac_vst_all <-  phyloseq::ordinate(vst_all, method="PCoA", distance = "wunifrac")

PCoA_wunifrac_vst_all_plot  <-
  plot_ordination(vst_all , PCoA_wunifrac_vst_all, color = "spike", shape ="sample_time", title = "wUnifrac PCoA") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors)

PCoA_wunifrac_vst_all_plot


```

```{r wunifrac-just-t6}

wunifrac.t6 <- ordinate(vst_all.t6, method="PCoA", distance = "wunifrac")

wunifrac.t6.plot  <-
  plot_ordination(vst_all.t6, wunifrac.t6, color = "treatment", title = "")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw() +
    theme(legend.title = element_blank())+
    scale_shape_discrete(na.translate = F) #+
   # scale_color_manual(values = c("black", "#3DB7E9",  "#D55E00", "#F0E442"), na.value = NA, labels = c("None", "Antibiotics", "Fe(OH)3", "Na2SO4", " "))

wunifrac.t6.plot #print the plot


ggsave("figures/wunifrac.vst_all_t6.jpeg", height = 6 )


```



```{r pcoa-wunifrac-vst-all-treatment}

PCoA_wunifrac_vst_all_plot  <-
  plot_ordination(vst_all , PCoA_wunifrac_vst_all, color = "treatment", shape ="sample_time", title = "wUnifrac PCoA") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors) + 
  theme_bw()

PCoA_wunifrac_vst_all_plot +  theme(legend.title = element_blank())

ggsave("figures/weightedunifrac_PCOA-vst-live.jpeg", height = 6)

```


This part doesn't work because there are too few points
```{r draw-ellipse}
# 
# PCoA_wunifrac_vst_all_plot +
#   scale_color_manual(values= exper_colors) +
#   geom_point(size=3.5) +
#  stat_ellipse(type="norm", geom="polygon", alpha=1/10)


```



### Unweighted Unifrac
This only works with untransformed data; the original phyloseq object (ps_all)
```{r pcoa-unifrac-untransformed}


PCoA_unifrac_vst_all <-  ordinate(ps_all, method="PCoA", distance = "unifrac")

PCoA_unifrac_vst_all_plot <-  plot_ordination(ps_all, PCoA_unifrac_vst_all, color = "treatment", shape ="sample_time", title = "unifrac PCoA Untransformed") +
    scale_color_manual(values= exper_colors) +
    geom_point(size=3.5)

PCoA_unifrac_vst_all_plot + 
  theme_bw()+
  theme(legend.title = element_blank())


ggsave("figures/unifracPCOA-vst-live.jpeg", height = 6)


```


## Killed Samples

The killed samples have an extra data time sample @ T6 because we didn't take the T6 sample for the live samples
```{r remove-t7, results = 'hide'}

ps_killed.rmt7 <- subset_samples(ps_killed, sample_time != "T6")
vst_killed.rmt7 <-subset_samples(vst_killed, sample_time != "T6")

as.character(sample_data(ps_killed.rmt7)$sample_time)

#now we need to rename T7 to T6

samdata_rmt7 <- ps_killed.rmt7@sam_data

samdata_rmt7 <- samdata_rmt7 %>% 
  mutate(sample_time = case_when(sample_time == "T7" ~ "T6", 
                                 TRUE ~ sample_time))
         

as.factor(samdata_rmt7$sample_time) #change back to a factor 




```


```{r update-sampledata-killed}

sample_data(ps_killed.rmt7) = samdata_rmt7
sample_data(vst_killed.rmt7) = samdata_rmt7

```


### Bray Curtis 
Here we're using the new treatment variable we made to make the plot
```{r nmds-bray-vst-killed-treatment}

nmds_bray_vst_killed <- ordinate(vst_killed, method="NMDS", distance = "bray")



nmds_bray_vst_killed.plot  <-
  plot_ordination(vst_killed, nmds_bray_vst_killed, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray_vst_killed.plot #print the plot


```

```{r nmds-bray-vst-killed-treatment.rmt7}

nmds_bray_vst_killed.rmt7 <- ordinate(vst_killed.rmt7, method="NMDS", distance = "bray")



nmds_bray_vst_killed.plot.rmt7  <-
  plot_ordination(vst_killed.rmt7, nmds_bray_vst_killed.rmt7, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS killed")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw() +
    theme(legend.title = element_blank())


nmds_bray_vst_killed.plot.rmt7 #print the plot

ggsave("figures/BrayNMDS-killed-vst.jpeg", height = 6)


```




### Weighted Unifrac 
```{r pcoa-wunifrac-vst-killed}

PCoA_wunifrac_vst_killed <-  phyloseq::ordinate(vst_killed.rmt7, method="PCoA", distance = "wunifrac")

PCoA_wunifrac_vst_all_killed.plot  <-
  plot_ordination(vst_killed.rmt7 , PCoA_wunifrac_vst_killed, color = "treatment", shape ="sample_time", title = "wUnifrac PCoA killed") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors)+
  theme_bw()+
  theme(legend.title = element_blank())

PCoA_wunifrac_vst_all_killed.plot

ggsave("figures/weightedunifrace_pcoa-killed-vst.jpeg", height = 6)


```

### unweighted unifrace
```{r pcoa_unifrac-ps_killed}

PCoA_unifrac_ps_killed <-  phyloseq::ordinate(ps_killed.rmt7, method="PCoA", distance = "unifrac")

PCoA_unifrac_ps_killed.plot  <-
  plot_ordination(ps_killed.rmt7 , PCoA_unifrac_ps_killed , color = "treatment", shape ="sample_time", title = "Unifrac PCoA killed") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors)+
  theme_bw()+
  theme(legend.title = element_blank())

PCoA_unifrac_ps_killed.plot

ggsave("figures/unifrace_pcoa-killed-vst.jpeg", height = 6)



```




## Combine Killed and Live samples vst 

For this I would likely have to either make a phloseq object with both sets of samples from the beginning, or make 2 objects without the trees, and then merge them together. Honestly not sure it's worth comparing these two datasets; might come back to it later. 

```{r vst-combined-metadata}

#load no tree phyloseq objects 
vst_all.nt <- readRDS("~/landfill-microcosms/data/live_samples_objects/vst_all.nt")
vst_killed.nt <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.nt")

#combine phyloseq objects
vst_combo <- merge_phyloseq(vst_all.nt, vst_killed.nt)

#extract metadata 
meta_combo <- sample_data(vst_combo)

#add a variable that identifies which come from the live and which come from the killed, for some reason, the sample column in the killed metadata is empty, so we can use that to your advantage 
meta_combo <- meta_combo %>% 
  mutate(experiment = case_when(is.na(sample) ~ "killed", 
                            TRUE ~ "live"))


sample_data(vst_combo) <- meta_combo #reassign the new metadata table to the sample_data of the merged phyloseq object

```

```{r vst-combo-nmds-bray}


nmds_bray.vst_combo <- ordinate(vst_combo, method="NMDS", distance = "bray")



nmds_bray.vst_combo.plot  <-
  plot_ordination(vst_combo, nmds_bray.vst_combo, color = "experiment", title = "Bray-Curtis NMDS")+ 
    geom_point(size=3.5)+ 
    theme_bw()




nmds_bray.vst_combo.plot

```
The NMDS plot looks super weird, presumably because I've used the transformed data and they are normalized internally somehow

I'm going to use the untransformed data instead. 

What I should really do is run through the dada2 piple with all the samples in one group, but I'm guess that would take FOREVER to run on this computer, so we're going to this work around instead. 
```{r ps-combined-metadata}


ps_all.nt <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.nt")

ps_killed.nt <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.nt")


ps_combo <- merge_phyloseq(ps_all.nt, ps_killed.nt)

meta_combo.ps <- sample_data(ps_combo)

meta_combo.ps <- meta_combo.ps %>% 
  mutate(experiment = case_when(is.na(sample) ~ "killed", #when sample column is NA, assign the new variable to killed 
                            TRUE ~ "live")) #otherwise, assign it live


sample_data(ps_combo) <- meta_combo.ps

```

```{r ps-combo-nmds-bray}


nmds_bray.ps_combo <- ordinate(ps_combo, method="NMDS", distance = "bray")



nmds_bray.ps_combo.plot  <-
  plot_ordination(ps_combo, nmds_bray.ps_combo, color = "experiment", title = "Bray-Curtis NMDS")+ 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray.ps_combo.plot

```
This looks much better, but I think I still need to remove the T6 from the killed samples because there are far too many points for killed. 

```{r killed.rmt7}


ps_killed.nt.rmt7 <- subset_samples(ps_killed.nt, sample_time != "T6") #remove T6 from the no tree ps object 

#now we need to rename T7 to T6

samdata.nt.rmt7 <- ps_killed.nt.rmt7@sam_data

samdata.nt.rmt7 <- samdata.nt.rmt7 %>% 
  mutate(sample_time = case_when(sample_time == "T7" ~ "T6", 
                                 TRUE ~ sample_time))
         

as.factor(samdata.nt.rmt7$sample_time) #change back to a factor, should now be T1-T6 

sample_data(ps_killed.nt.rmt7) <- samdata.nt.rmt7

#remake the ps object w T7 removed
ps_combo.rmt7 <- merge_phyloseq(ps_all.nt, ps_killed.nt.rmt7)

meta_combo.ps.rmt7 <- sample_data(ps_combo.rmt7)

meta_combo.ps.rmt7 <- meta_combo.ps.rmt7 %>% 
  mutate(experiment = case_when(is.na(sample) ~ "killed", #when sample column is NA, assign the new variable to killed 
                            TRUE ~ "live")) #otherwise, assign it live


sample_data(ps_combo.rmt7) <- meta_combo.ps.rmt7



```

```{r ps-combo-nmds-bray-rmt7}


nmds_bray.ps_combo.rmt7 <- ordinate(ps_combo.rmt7, method="NMDS", distance = "bray")


nmds_bray.ps_combo.plot.rmt7  <-
  plot_ordination(ps_combo.rmt7, nmds_bray.ps_combo.rmt7, color = "experiment", title = "Bray-Curtis NMDS")+ 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray.ps_combo.plot.rmt7

```

https://colorbrewer2.org/#type=sequential&scheme=YlGnBu&n=6
```{r nicer-plot-ps-combo-nmds-bray-rmt7}

cols = c("#5A000F", "#00CBA7")

cols_disc = c('#1b9e77','#d95f02','#7570b3','#e7298a','#66a61e','#e6ab02')

cols_pink <- c('#feebe2','#fcc5c0','#fa9fb5','#f768a1','#c51b8a','#7a0177')

cols_blue <- c('#FEDC56','#c7e9b4','#7fcdbb','#41b6c4','#2c7fb8','#253494')

nmds_bray.ps_combo.plot.rmt7  <-
  plot_ordination(ps_combo.rmt7, nmds_bray.ps_combo.rmt7, shape = "experiment", title = "Bray-Curtis NMDS" , color = "sample_time")+ 
    geom_point(size=3)+ 
    theme_bw() +
  scale_color_manual(values = cols_blue)+
  scale_shape_manual(values= c(15 ,19))+
  theme(legend.title = element_blank()) 

  # stat_ellipse(aes(ps_combo.rmt7, shape = "experiment"), type="norm", geom="polygon", alpha=1/10)
  
nmds_bray.ps_combo.plot.rmt7


ggsave("figures/combined-nmds-bray.jpeg", height = 6, width = 7)

#different shapes

nmds_bray.ps_combo.plot.rmt7  <-
  plot_ordination(ps_combo.rmt7, nmds_bray.ps_combo.rmt7, shape = "experiment", title = "Bray-Curtis NMDS" , color = "sample_time")+ 
    geom_point(size=3)+ 
    theme_bw() +
  scale_color_manual(values = cols_blue)+
  scale_shape_manual(values= c(17 ,19))+
  theme(legend.title = element_blank()) 

  # stat_ellipse(aes(ps_combo.rmt7, shape = "experiment"), type="norm", geom="polygon", alpha=1/10)
  
nmds_bray.ps_combo.plot.rmt7


ggsave("figures/combined-nmds-bray2.jpeg", height = 6, width = 7)


```


```{r check-taxtab, eval = F}

ps_all.nt.taxtab = tax_table(ps_all.nt)

check_df_all <- as.data.frame(ps_all.nt.taxtab)

ps_killed.nt.rmt7.taxtab = tax_table(ps_killed.nt.rmt7)

check_df_killed <- as.data.frame(ps_killed.nt.rmt7.taxtab)


```




# Group Signifcance Testing with ADONIS 
another way:
https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/beta-diversity-metrics.html#phylogenetic-beta-diversity-metrics


paper source:
https://onlinelibrary.wiley.com/doi/10.1002/9781118445112.stat07841

```{r adonisfunction}

#Set a random seed so that exact results can be reproduced
set.seed(10000)


#function to run ADONIS test on a phyloseq object and a variable from metadata 
doadonis_wunifrac <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "wunifrac")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col, permutations = 10000) 
  #the great number of permuations gives greater precision, but also takes longer to run, although this didn't take that long
  print("Adonis results:")
  print(adonis.bdist)}


doadonis_bray <- function(physeq, category) {
  bdist <- phyloseq::distance(physeq, "bray")
  col <- as(sample_data(physeq), "data.frame")[ ,category]
  # Adonis test
  adonis.bdist <- adonis(bdist ~ col, permutations = 10000) 
  #the great number of permuations gives greater precision, but also takes longer to run, although this didn't take that long
  print("Adonis results:")
  print(adonis.bdist)}


```


```{r adonis-all-timepoints}

all_timepoints_adonis = doadonis_wunifrac(vst_all, "treatment") #Results are significant, though I'm not sure which groups are different from each other. 

all_timepoints_adonis$aov.tab$'Pr(>F)'[1] #this gives you the p value of the adonis test 
#same thing but in another way; double bracket notation; 
all_timepoints_adonis[["aov.tab"]][["Pr(>F)"]][1]

```


```{r adonist1-t4}

ps_all.t4 <- ps_all %>% subset_samples(sample_time %in% c("T1", "T2", "T3", "T4"))

adonis.t4.wu <- doadonis_wunifrac(ps_all.t4, "sample_time") #significant, 

#but how do we find out which group is different from the other?


```

```{r check-homogenity-t4}

ps_all.t4_unifrac.dis <- UniFrac(ps_all.t4)

ps_all.t4_meta <- sample_data(ps_all.t4)

ps.disper <- betadisper(ps_all.t4_unifrac.dis, ps_all.t4_meta$sample_time)

permutest(ps.disper, pairwise = TRUE)

plot(ps.disper)

boxplot(ps.disper)

TukeyHSD(ps.disper)



anova(ps.disper)

```
Not sigificant but idk if that implies homogenity or not homogenity 

https://rfunctions.blogspot.com/2019/03/betadisper-and-adonis-homogeneity-of.html

Based on the blog, we meet the adonis condition of homogenity P<0.105
 
```{r t2-t3-t4}

ps_all.ntt1 <- ps_all%>%
  subset_samples(sample_time %in% c("T2", "T3", "T4"))

doadonis_bray(ps_all.ntt1, "sample_time")

set.seed(1000)
adonis.ntt1.wu <- doadonis_wunifrac(ps_all.ntt1, "sample_time")

str(adonis.ntt1.wu)

```




# Alpha diversity 
Used to identify within individual taxa richness and evenness
The commonly used metrics/indices are Shannon, Inverse Simpson, Simpson, Gini, Observed and Chao1. These indices do not take into account the phylogeny of the taxa identified in sequencing. Phylogenetic diversity (Faithâ€™s PD) uses phylogenetic distance to calculate the diversity of a given sample. (Sudarshan et al., 2020)


One has to consider the sequencing depth (how much of the taxa have been sampled) for each sample. If there is a large difference, then it is important to normalize the samples to equal sampling depth. First, we look at the sampling depth (no. of reads per sample).
```{r}

summary(sample_sums(ps_all))

#we can plot the rarefaction curve for the observed ASVs in the entire dataset, this is a way to check how the richness captured in the sequencing effort 

otu_tab = t(abundances(ps_all))  #t = transpose I think, not sure why we transposed it



#plot of species/reads
#we're nto doing anything with this, but it's good to know
p = vegan::rarecurve(otu_tab, 
                      step = 50, label = FALSE, 
                      sample = min(rowSums(otu_tab), 
                                   col = "blue", cex = 0.6))

#the tutorial I'm using here used this plot to rarify their data, however, we're not doing this here because susan holmes would be mad.

```

Non-phylogenetic diversities (non vst)
```{r create-table-of-diversity-metrics}
#make sure you have the most up-to-date version of the microbiome package; use #BiocManager::install("microbiome") #update all
#sometimes you have to unlaod and re-load the package for this step to work 
#let's calculate diversity
#this also doesn't work with vst_all
#turns out it's not suppose to work with the variance stabilizing transformation, so just use the original phyloseq object
all <- ps_all #we'll be working with this phyloseq object from here

div = alpha(all, index="all")  

data.table(div)

# get the metadata out as seprate object
meta <- meta(all)

# Add the rownames as a new colum for easy integration later.
meta$sam_name <- rownames(meta)

# Add the rownames to diversity table
div$sam_name <- rownames(div)

# merge these two data frames into one
div.df <- merge(div, meta, by = "sam_name")

# check the tables
div.df


```

```{r plot-shannon}

# 
# p_div <- ggboxplot(div.df, 
#                x = "spike", 
#                y = "diversity_shannon",
#               fill = "spike", 
#               palette = "jco")
# 
# p_div <- p_div + rotate_x_text()
# 
# print(p_div)
# 
# 
# 
# #Seperate by both spike and time using metadata
# p_div_1 <- ggboxplot(div.df, 
#                x = "spike_time", 
#                y = "diversity_shannon",
#               fill = "sample_time")
# 
# 
# p_div_1 <- p_div_1 + rotate_x_text()
# 
# print(p_div_1)


```

```{r re-level-samples}

as.factor(div.df$spike_time)

div.df$spike_time = factor(div.df$spike_time, levels = c("ABT1", "ABT2", "ABT3", "ABT4" , "ABT5" ,"ABT6", "FeT1","FeT2" ,"FeT3" ,"FeT4", "FeT5" ,"FeT6", "NLT1", "NLT2", "NLT3", "NLT4", "NLT5", "NLT6", "ST1" , "ST2",  "ST3",  "ST4", "ST5" , "ST6"))


```

shannon phylogenetic diversity w/re-ordered samples (non vst)
```{r plot-shannon-2}
#Seperate by both spike and time using metadata



p_div_1 <- ggboxplot(div.df, 
               x = "spike_time", 
               y = "diversity_shannon",
              fill = "sample_time")


p_div_1 <- p_div_1 + rotate_x_text()

print(p_div_1)

#THAT WORKED, IT'S ORDERED NOW!! ONLY TOOK 1 HOUR TO FIGURE OUT. 

```

```{r faceted-shannon-plot}

jpeg("~/landfill-microcosms/step-4-ordination_diversitymetrics_statistics/Figures/faceted-shannon-plot.jpeg", quality = 100)


div.df %>%
  group_by(spike, sample_time, treatment) %>% 
  dplyr::summarise(avg_shan = mean(diversity_shannon)) %>% 
  ggplot(aes(sample_time, avg_shan, color = treatment)) + 
  geom_point()+ 
  facet_wrap(~spike)+
  theme_bw()+
  scale_color_manual(values = exper_colors)

dev.off()
  



```


simpson boxplots (non vst)
```{r}
#spike time
p_even <- ggboxplot(div.df, 
               x = "spike_time", 
               y = "evenness_simpson",
              fill = "sample_time")


p_even <- p_even + rotate_x_text()

print(p_even)

```



Shannon and Simpson 
```{r diversity-with-phyloseq-package}

Shannon_plot <- 
  plot_richness(ps_all, x="sample", measures=c("Shannon"), color="spike")

Shannon_plot + scale_color_manual(values= exper_colors) 

Simpson_plot = plot_richness(ps_all, x="spike_time", measures=c("Simpson"), color="spike")

Simpson_plot  + scale_color_manual(values= exper_colors)
```

taxa glom
```{r}
# #summes all ASVs that were in the same level
# 
# #Glom at famiily level 
# vst_all_rel_glom = tax_glom(vst_all_rel_1, taxrank = "Family")
# vst_all_rel_glom_family = vst_all_rel_glom
# 
# #Glom at phylum level 
# vst_all_rel_glom_phylum = tax_glom(vst_all_rel_1, taxrank= "Phylum")
# 


```

```{r save-physeq-objects-madehere}

saveRDS(ps_all, "~/landfill-microcosms/data/live_samples_objects/ps_all.step4")
saveRDS(vst_all, "~/landfill-microcosms/data/live_samples_objects/vst_all.step4")


saveRDS(ps_killed.rmt7, "~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rmt7")
saveRDS(vst_killed.rmt7, "~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.rmt7")

saveRDS(ps_killed, "~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.step4")
saveRDS(vst_killed, "~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.step4")


```

