---
title: "ordination & diversity metrics"
author: "Judy-Malas"
date: "12/22/2021"
output: html_document
---

# Before Starting 

* Make sure your clear the global environment by clicking on the broom in R studio 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
  
### Load packages
```{r load-packages, message=FALSE, warning=FALSE}

library(ggplot2)
library(gridExtra)
library(phyloseq)
require("tidyverse")
library("dplyr")
library("rmarkdown")
library("vegan")
library("PoiClaClu")
library("doParallel")
library("plotly")
library("microbiome")
library("ggpubr")
library("viridis")
library("plyr")
library("magrittr")
library("scales")
library(microbiome) # data analysis and visualisation
library(phyloseq) # also the basis of data object. Data analysis and visualisation
library(RColorBrewer) # nice color options
library(ggpubr) # publication quality figures, based on ggplot2
library(DT) # interactive tables in html and markdown
library(dplyr) # data handling  
library(data.table)


```


```{r import-all-objects}

# live samples 

ps_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.step2") 

ps_all.rmout <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rmout") #outlier removed

vst_all.step2 <- readRDS("~/landfill-microcosms/data/live_samples_objects/vst_all.step2")

ps_all.rl <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.rl") 

ps_all.prop <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.prop") 

ps_all.5 <- readRDS("~/landfill-microcosms/data/live_samples_objects/ps_all.5") #low prevelance taxa removed, untransformed



 #killed samples 

ps_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.step2")

ps_killed.rmout <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rmout")

vst_killed.step2 <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.step2")

ps_killed.rl <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rl")

ps_killed.prop <- readRDS("~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.prop")


```



# Update Phyloseq Objects
#### update the metadata tables

```{r add_new_column}

ps_all.step2_samdata <- ps_all.step2@sam_data #extract the sample data 

new_sam_data <- ps_all.step2_samdata %>% 
  mutate(treatment =  #add a new column of data which specifies whether the samples have ACTUALLY been treated with a spike
           case_when(sample_time == "T1" ~ "None", 
                     sample_time == "T2" ~ "None", 
                    sample_time == "T3" ~ "None", 
                        sample_time == "T4" ~ "None", 
                        sample_time == "T5" ~ as.character(spike), 
                    sample_time == "T6" ~ as.character(spike))) 

new_sam_data <- new_sam_data %>% 
  mutate(treatment = ifelse(treatment == "Control", "None", treatment)) #change "Control" to "none" in the new treatment variable


as.factor(new_sam_data$treatment)

new_sam_data <- new_sam_data %>% 
  mutate(treatment = fct_relevel(treatment, "None", "Antibiotics", "Fe(OH)3", "Na2SO4"))
  


```

```{r add_new_column_killed_samdata}

ps_killed.step2_samdata <- ps_killed.step2@sam_data #extract the sample data 

new_sam_data_killed <- ps_killed.step2_samdata %>% 
  mutate(treatment =  #add a new column of data which specifies whether the samples have ACTUALLY been treated with a spike
           case_when(sample_time == "T1" ~ "None", 
                     sample_time == "T2" ~ "None", 
                    sample_time == "T3" ~ "None", 
                        sample_time == "T4" ~ "None", 
                        sample_time == "T5" ~ as.character(spike), 
                    sample_time == "T6" ~ as.character(spike), 
                    sample_time == "T7"~as.character(spike))) 

new_sam_data_killed <- new_sam_data_killed %>% 
  mutate(treatment = ifelse(treatment == "killed_control", "None", treatment)) #change "Control" to "none" in the new treatment variable


as.factor(new_sam_data_killed$treatment)

new_sam_data_killed <- new_sam_data_killed %>% 
  mutate(treatment = fct_relevel(treatment, "None", "Antibiotics", "Fe(OH)3", "Na2SO4"))

levels(new_sam_data_killed$treatment)

```


```{r update-physeq-objects}

ps_all.step2@sam_data = new_sam_data # assign the new dataframe to the phyloseq object
vst_all.step2@sam_data = new_sam_data
ps_killed.step2@sam_data = new_sam_data_killed
vst_killed.step2@sam_data = new_sam_data_killed


```



#### re-simplifying names
```{r rename-physeq-objects}

ps_all <- ps_all.step2
vst_all <- vst_all.step2
ps_killed <- ps_killed.step2
vst_killed <- vst_killed.step2

```



####  Assigning plot colors


These are the original colors I was using, but they are not colorblind friendly
```{r assign-colors}

# exper_colors <-  c( "gray50", "blue", "firebrick1", "gold")

```

```{r list-all-cb-friendly-palettes}

brewer.pal.info %>% 
  filter(colorblind == T)
  
```

```{r visualize-all-cb-friendly-palettes}
display.brewer.all(colorblindFriendly = T)
```


Ended up using colors in the 8-color plalette for color blindness: 

http://mkweb.bcgsc.ca/colorblind/img/colorblindness.palettes.v11.pdf

```{r updated-assign-colors}
exper_colors <- c("black", "#3DB7E9",  "#D55E00", "#F0E442")
```


# Distance Measures; Ordination plots

## Live Samples 

### Bray-Curtis
```{r nmds-bray-vst-all}

nmds_bray_vst_all <- ordinate(vst_all, method="NMDS", distance = "bray")

nmds_bray_vst_all_plot  <-
  plot_ordination(vst_all, nmds_bray_vst_all, color = "spike", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)
nmds_bray_vst_all_plot #print the plot


```

Here we're using the new treatment variable we made to make the plot
```{r nmds-bray-vst-all-treatment}

nmds_bray_vst_all_plot  <-
  plot_ordination(vst_all, nmds_bray_vst_all, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray_vst_all_plot #print the plot


```


### Weighted Unifrac 
```{r pcoa-wunifrac-vst-all}

PCoA_wunifrac_vst_all <-  phyloseq::ordinate(vst_all, method="PCoA", distance = "wunifrac")

PCoA_wunifrac_vst_all_plot  <-
  plot_ordination(vst_all , PCoA_wunifrac_vst_all, color = "spike", shape ="sample_time", title = "wUnifrac PCoA") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors)

PCoA_wunifrac_vst_all_plot


```

```{r pcoa-wunifrac-vst-all-treatment}

PCoA_wunifrac_vst_all_plot  <-
  plot_ordination(vst_all , PCoA_wunifrac_vst_all, color = "treatment", shape ="sample_time", title = "wUnifrac PCoA") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors) + 
  theme_bw()

PCoA_wunifrac_vst_all_plot

```


This part doesn't work because there are too few points
```{r draw-ellipse}

# PCoA_wunifrac_vst_all_plot +
#   scale_color_manual(values= exper_colors) +
#   geom_point(size=3.5) +
#   stat_ellipse(type="norm", geom="polygon", alpha=1/10)


```



### Unweighted Unifrac
This only works with untransformed data; the original phyloseq object (ps_all)
```{r pcoa-unifrac-untransformed}


PCoA_unifrac_vst_all <-  ordinate(ps_all, method="PCoA", distance = "unifrac")

PCoA_unifrac_vst_all_plot <-  plot_ordination(ps_all, PCoA_unifrac_vst_all, color = "treatment", shape ="sample_time", title = "unifrac PCoA Untransformed") +
    scale_color_manual(values= exper_colors) +
    geom_point(size=3.5)

PCoA_unifrac_vst_all_plot


```


## Killed Samples

The killed samples have an extra data time sample @ T6 because we didn't take the T6 sample for the live samples
```{r remove-t7}

ps_killed.rmt7 <- subset_samples(ps_killed, sample_time != "T6")
vst_killed.rmt7 <-subset_samples(vst_killed, sample_time != "T6")

as.character(sample_data(ps_killed.rmt7)$sample_time)

#now we need to rename T7 to T6

samdata_rmt7 <- ps_killed.rmt7@sam_data

samdata_rmt7 <- samdata_rmt7 %>% 
  mutate(sample_time = case_when(sample_time == "T7" ~ "T6", 
                                 TRUE ~ sample_time))
         

as.factor(samdata_rmt7$sample_time) #change back to a factor 




```


```{r update-sampledata-killed}

sample_data(ps_killed.rmt7) = samdata_rmt7
sample_data(vst_killed.rmt7) = samdata_rmt7

```


### Bray Curtis 
Here we're using the new treatment variable we made to make the plot
```{r nmds-bray-vst-killed-treatment}

nmds_bray_vst_killed <- ordinate(vst_killed, method="NMDS", distance = "bray")



nmds_bray_vst_killed.plot  <-
  plot_ordination(vst_killed, nmds_bray_vst_killed, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray_vst_killed.plot #print the plot


```

```{r nmds-bray-vst-killed-treatment.rmt7}

nmds_bray_vst_killed.rmt7 <- ordinate(vst_killed.rmt7, method="NMDS", distance = "bray")



nmds_bray_vst_killed.plot.rmt7  <-
  plot_ordination(vst_killed.rmt7, nmds_bray_vst_killed.rmt7, color = "treatment", shape ="sample_time", title = "Bray-Curtis NMDS")+ 
    scale_color_manual(values= exper_colors) + 
    geom_point(size=3.5)+ 
    theme_bw()


nmds_bray_vst_killed.plot.rmt7 #print the plot


```


### Weighted Unifrac 
```{r pcoa-wunifrac-vst-killed}

PCoA_wunifrac_vst_killed <-  phyloseq::ordinate(vst_killed.rmt7, method="PCoA", distance = "wunifrac")

PCoA_wunifrac_vst_all_killed.plot  <-
  plot_ordination(vst_killed.rmt7 , PCoA_wunifrac_vst_killed, color = "treatment", shape ="sample_time", title = "wUnifrac PCoA killed") +
  geom_point(size=3.5) +
  scale_color_manual(values= exper_colors)

PCoA_wunifrac_vst_all_killed.plot


```


##### Combine Killed and Live samples vst 

For this I would likely have to either make a phloseq object with both sets of samples from the beginning, or make 2 objects without the trees, and then merge them together. Honestly not sure it's worth comparing these two datasets; might come back to it later. 
```{r}

#vst_combo <- merge_phyloseq(vst_all, vst_killed.rmt7)
## Error in FUN(X[[i]], ...) : one tree has a different number of tips


```



# Alpha diversity 
Used to identify within individual taxa richness and evenness
The commonly used metrics/indices are Shannon, Inverse Simpson, Simpson, Gini, Observed and Chao1. These indices do not take into account the phylogeny of the taxa identified in sequencing. Phylogenetic diversity (Faithâ€™s PD) uses phylogenetic distance to calculate the diversity of a given sample. (Sudarshan et al., 2020)


One has to consider the sequencing depth (how much of the taxa have been sampled) for each sample. If there is a large difference, then it is important to normalize the samples to equal sampling depth. First, we look at the sampling depth (no. of reads per sample).
```{r}

summary(sample_sums(ps_all))

#we can plot the rarefaction curve for the observed ASVs in the entire dataset, this is a way to check how the richness captured in the sequencing effort 

otu_tab = t(abundances(ps_all))  #t = transpose I think, not sure why we transposed it



#plot of species/reads
#we're nto doing anything with this, but it's good to know
p = vegan::rarecurve(otu_tab, 
                      step = 50, label = FALSE, 
                      sample = min(rowSums(otu_tab), 
                                   col = "blue", cex = 0.6))

#the tutorial I'm using here used this plot to rarify their data, however, we're not doing this here because susan holmes would be mad.

```

Non-phylogenetic diversities (non vst)
```{r create-table-of-diversity-metrics}
#make sure you have the most up-to-date version of the microbiome package; use #BiocManager::install("microbiome") #update all
#sometimes you have to unlaod and re-load the package for this step to work 
#let's calculate diversity
#this also doesn't work with vst_all
#turns out it's not suppose to work with the variance stabilizing transformation, so just use the original phyloseq object
all <- ps_all #we'll be working with this phyloseq object from here

div = alpha(all, index="all")  

data.table(div)

# get the metadata out as seprate object
meta <- meta(all)

# Add the rownames as a new colum for easy integration later.
meta$sam_name <- rownames(meta)

# Add the rownames to diversity table
div$sam_name <- rownames(div)

# merge these two data frames into one
div.df <- merge(div, meta, by = "sam_name")

# check the tables
div.df


```

```{r plot-shannon}

# 
# p_div <- ggboxplot(div.df, 
#                x = "spike", 
#                y = "diversity_shannon",
#               fill = "spike", 
#               palette = "jco")
# 
# p_div <- p_div + rotate_x_text()
# 
# print(p_div)
# 
# 
# 
# #Seperate by both spike and time using metadata
# p_div_1 <- ggboxplot(div.df, 
#                x = "spike_time", 
#                y = "diversity_shannon",
#               fill = "sample_time")
# 
# 
# p_div_1 <- p_div_1 + rotate_x_text()
# 
# print(p_div_1)


```

```{r re-level-samples}

as.factor(div.df$spike_time)

div.df$spike_time = factor(div.df$spike_time, levels = c("ABT1", "ABT2", "ABT3", "ABT4" , "ABT5" ,"ABT6", "FeT1","FeT2" ,"FeT3" ,"FeT4", "FeT5" ,"FeT6", "NLT1", "NLT2", "NLT3", "NLT4", "NLT5", "NLT6", "ST1" , "ST2",  "ST3",  "ST4", "ST5" , "ST6"))


```


shannon phylogenetic diversity w/re-ordered samples (non vst)
```{r plot-shannon}
#Seperate by both spike and time using metadata



p_div_1 <- ggboxplot(div.df, 
               x = "spike_time", 
               y = "diversity_shannon",
              fill = "sample_time")


p_div_1 <- p_div_1 + rotate_x_text()

print(p_div_1)

#THAT WORKED, IT'S ORDERED NOW!! ONLY TOOK 1 HOUR TO FIGURE OUT. 

```

```{r faceted-shannon-plot}

jpeg("~/landfill-microcosms/step-4-ordination_diversitymetrics_statistics/Figures/faceted-shannon-plot.jpeg", quality = 100)


div.df %>%
  group_by(spike, sample_time, treatment) %>% 
  dplyr::summarise(avg_shan = mean(diversity_shannon)) %>% 
  ggplot(aes(sample_time, avg_shan, color = treatment)) + 
  geom_point()+ 
  facet_wrap(~spike)+
  theme_bw()+
  scale_color_manual(values = exper_colors)

dev.off()
  



```


simpson boxplots (non vst)
```{r}
#spike time
p_even <- ggboxplot(div.df, 
               x = "spike_time", 
               y = "evenness_simpson",
              fill = "sample_time")


p_even <- p_even + rotate_x_text()

print(p_even)

```



Shannon and Simpson 
```{r diversity-with-phyloseq-package}

Shannon_plot <- 
  plot_richness(ps_all, x="sample", measures=c("Shannon"), color="spike")

Shannon_plot + scale_color_manual(values= exper_colors) 

Simpson_plot = plot_richness(ps_all, x="spike_time", measures=c("Simpson"), color="spike")

Simpson_plot  + scale_color_manual(values= exper_colors)
```


taxa glom
```{r}
# #summes all ASVs that were in the same level
# 
# #Glom at famiily level 
# vst_all_rel_glom = tax_glom(vst_all_rel_1, taxrank = "Family")
# vst_all_rel_glom_family = vst_all_rel_glom
# 
# #Glom at phylum level 
# vst_all_rel_glom_phylum = tax_glom(vst_all_rel_1, taxrank= "Phylum")
# 


```

```{r save-physeq-objects-madehere}

saveRDS(ps_all, "~/landfill-microcosms/data/live_samples_objects/ps_all.step4")
saveRDS(vst_all, "~/landfill-microcosms/data/live_samples_objects/vst_all.step4")


saveRDS(ps_killed.rmt7, "~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.rmt7")
saveRDS(vst_killed.rmt7, "~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.rmt7")

saveRDS(ps_killed, "~/landfill-microcosms/data/killed_microcosms_objects/ps_killed.step4")
saveRDS(vst_killed, "~/landfill-microcosms/data/killed_microcosms_objects/vst_killed.step4")


```

